suite/tests/client-interface/drx-scattergather-aarch64.cpp

```diff
@@ -0,0 +1,1170 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Arm Limited nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL ARM LIMITED OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include <array>
+#include <cassert>
+#include <cstdint>
+#include <cstring>
+#include <functional>
+#include <map>
+#include <string>
+#include <sys/prctl.h>
+#include <vector>
+
+#include "tools.h"
+
+namespace {
+
+/*
+ * Tests are specified assuming 128-bit vectors. If we run on hardware with a
+ * higher VL then vector values are made up to the correct size by duplicating
+ * the first 128-bits.
+ */
+constexpr size_t TEST_VL_BYTES = 16;
+
+constexpr size_t NUM_Z_REGS = 32;
+constexpr size_t NUM_P_REGS = 16;
+
+using vector_reg_value128_t = std::array<uint8_t, TEST_VL_BYTES>;
+using predicate_reg_value128_t = uint16_t;
+
+constexpr vector_reg_value128_t UNINITIALIZED_VECTOR {
+    0xAD, 0xDE, 0xAD, 0xDE, 0xAD, 0xDE, 0xAD, 0xDE,
+    0xAD, 0xDE, 0xAD, 0xDE, 0xAD, 0xDE, 0xAD, 0xDE,
+};
+constexpr predicate_reg_value128_t UNINITIALIZED_PREDICATE = 0xDEAD;
+
+enum class element_size_t {
+    BYTE = 1,
+    HALF = 2,
+    SINGLE = 4,
+    DOUBLE = 8,
+};
+
+/*! Exhaustive lists of valid 128-bit vl predicate register values for for
+ * single or double sized elements.
+ */
+const std::map<element_size_t, std::vector<predicate_reg_value128_t>> ALL_PREDICATES {
+    { element_size_t::SINGLE,
+      { 0x0000, 0x0001, 0x0010, 0x0011, 0x0100, 0x0101, 0x0110, 0x0111, 0x1000, 0x1001,
+        0x1010, 0x1011, 0x1100, 0x1101, 0x1110, 0x1111 } },
+    { element_size_t::DOUBLE, { 0x0000, 0x0001, 0x0100, 0x0101 } },
+};
+
+enum test_result_t {
+    FAIL,
+    PASS,
+};
+
+/*!
```

Comment:
OK, using `\param`, but match the `/**` style everywhere else in the codebase instead of `/*!`.  There is no pass that's running this through doxygen, so syntax errors or missing links or anything is not going to be caught.  Ditto below.